Vous êtes Data Scientist dans une start-up de la EdTech, nommée academy, qui propose des contenus de formation en ligne pour un public de niveau lycée et université.

Mark, votre manager, vous a convié à une réunion pour vous présenter le projet d’expansion à l’international de l’entreprise. Il vous confie une première mission d’analyse exploratoire, pour déterminer si les données sur l’éducation de la banque mondiale permettent d’informer le projet d’expansion.

Voici les différentes questions que Mark aimerait explorer, que vous avez notées durant la réunion :

Quels sont les pays avec un fort potentiel de clients pour nos services ?
Pour chacun de ces pays, quelle sera l’évolution de ce potentiel de clients ?
Dans quels pays l'entreprise doit-elle opérer en priorité ?
Votre mission
Mark vous a donc demandé de réaliser une analyse pré-exploratoire de ce jeu de données. Il vous a transmis cet email à la suite de la réunion :

Hello,

Les données de la Banque mondiale sont disponibles à l’adresse suivante :

https://datacatalog.worldbank.org/dataset/education-statistics

Ou en téléchargement direct à ce lien.

Je te laisse regarder la page d'accueil qui décrit le jeu de données. En résumé, l’organisme “EdStats All Indicator Query” de la Banque mondiale répertorie 4000 indicateurs internationaux décrivant l’accès à l’éducation, l’obtention de diplômes et des informations relatives aux professeurs, aux dépenses liées à l’éducation... Tu trouveras plus d'info sur ce site :

http://datatopics.worldbank.org/education/

Pour la pré-analyse, pourrais-tu :

Valider la qualité de ce jeu de données (comporte-t-il beaucoup de données manquantes, dupliquées ?)

Décrire les informations contenues dans le jeu de données (nombre de colonnes ? nombre de lignes ?)

Sélectionner les informations qui semblent pertinentes pour répondre à la problématique (quelles sont les colonnes contenant des informations qui peuvent être utiles pour répondre à la problématique de l’entreprise ?)

Déterminer des ordres de grandeurs des indicateurs statistiques classiques pour les différentes zones géographiques et pays du monde (moyenne/médiane/écart-type par pays et par continent ou bloc géographique)

Ton travail va nous permettre de déterminer si ce jeu de données peut informer les décisions d'ouverture vers de nouveaux pays. On va partager ton analyse avec le board, alors merci de soigner la présentation et de l'illustrer avec des graphiques pertinents et lisibles !

# CSV used : EdStatsData

## Libraries imports

In [204]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import os

## File imports

In [205]:
filename = 'EdStatsData.csv'
directory = '/CSV'
dfEdStatsData = pd.read_csv(os.getcwd() + directory + '/' + filename)

## Basic informations about the dataset

> Data sample :

In [206]:
dfEdStatsData.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975  ...  2060  \
0        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  ...   NaN   

   2065  2070  2075  2080  2085  2090  2095  2100  Unnamed: 69  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[5 rows x 70 columns]

> Data infos

In [207]:
dfEdStatsData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 886930 entries, 0 to 886929
Data columns (total 70 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    886930 non-null  object 
 1   Country Code    886930 non-null  object 
 2   Indicator Name  886930 non-null  object 
 3   Indicator Code  886930 non-null  object 
 4   1970            72288 non-null   float64
 5   1971            35537 non-null   float64
 6   1972            35619 non-null   float64
 7   1973            35545 non-null   float64
 8   1974            35730 non-null   float64
 9   1975            87306 non-null   float64
 10  1976            37483 non-null   float64
 11  1977            37574 non-null   float64
 12  1978            37576 non-null   float64
 13  1979            36809 non-null   float64
 14  1980            89122 non-null   float64
 15  1981            38777 non-null   float64
 16  1982            37511 non-null   float64
 17  1983      

Our dataset contains 886930 entries, we have 66 float variables and 4 object variables which are strings but we keep them in object type as they are handled better like this.


## Dataset fractioning

We create sub datasets for "country-only" and "worldwide"

In [208]:
# we convert the column names to serpent case
dfEdStatsData.columns = dfEdStatsData.columns.str.lower().str.replace(' ', '_')
dfEdStatsData

# world sub dataset
dfEdStatsDataWorld = dfEdStatsData[dfEdStatsData['country_name'] == 'World']

# country sub dataset
# we check for the first occurence of Afghanistan as it is the first country in the dataset
firstOccurence = dfEdStatsData[dfEdStatsData['country_name'] == 'Afghanistan'].index[0]
# we split the dataset into two parts
dfEdStatsDataCountry = dfEdStatsData.iloc[firstOccurence:]
dfEdStatsDataCountry.reset_index(drop=True, inplace=True)
dfEdStatsDataCountry

country_name country_code  \
0       Afghanistan          AFG   
1       Afghanistan          AFG   
2       Afghanistan          AFG   
3       Afghanistan          AFG   
4       Afghanistan          AFG   
...             ...          ...   
795300     Zimbabwe          ZWE   
795301     Zimbabwe          ZWE   
795302     Zimbabwe          ZWE   
795303     Zimbabwe          ZWE   
795304     Zimbabwe          ZWE   

                                           indicator_name  \
0       Adjusted net enrolment rate, lower secondary, ...   
1       Adjusted net enrolment rate, lower secondary, ...   
2       Adjusted net enrolment rate, lower secondary, ...   
3       Adjusted net enrolment rate, lower secondary, ...   
4       Adjusted net enrolment rate, primary, both sex...   
...                                                   ...   
795300  Youth illiterate population, 15-24 years, male...   
795301  Youth literacy rate, population 15-24 years, b...   
795302  Youth literacy rate, population 15-24 years, f...   
795303  Youth literacy rate, population 15-24 years, g...   
795304  Youth literacy rate, population 15-24 years, m...   

              indicator_code  1970  1971  1972  1973      1974  1975  ...  \
0                 UIS.NERA.2   NaN   NaN   NaN   NaN   7.05911   NaN  ...   
1               UIS.NERA.2.F   NaN   NaN   NaN   NaN   2.53138   NaN  ...   
2             UIS.NERA.2.GPI   NaN   NaN   NaN   NaN   0.22154   NaN  ...   
3               UIS.NERA.2.M   NaN   NaN   NaN   NaN  11.42652   NaN  ...   
4                SE.PRM.TENR   NaN   NaN   NaN   NaN       NaN   NaN  ...   
...                      ...   ...   ...   ...   ...       ...   ...  ...   
795300      UIS.LP.AG15T24.M   NaN   NaN   NaN   NaN       NaN   NaN  ...   
795301     SE.ADT.1524.LT.ZS   NaN   NaN   NaN   NaN       NaN   NaN  ...   
795302  SE.ADT.1524.LT.FE.ZS   NaN   NaN   NaN   NaN       NaN   NaN  ...   
795303  SE.ADT.1524.LT.FM.ZS   NaN   NaN   NaN   NaN       NaN   NaN  ...   
795304  SE.ADT.1524.LT.MA.ZS   NaN   NaN   NaN   NaN       NaN   NaN  ...   

        2060  2065  2070  2075  2080  2085  2090  2095  2100  unnamed:_69  
0        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
1        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
2        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
3        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
4        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
...      ...   ...   ...   ...   ...   ...   ...   ...   ...          ...  
795300   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
795301   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
795302   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
795303   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
795304   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[795305 rows x 70 columns]

## Null percentage by indicator

Ranking of most relevant indicators depending on null percentage for data aggregated **in the past**

In [209]:
dfNullPercentageByIndicator = dfEdStatsDataCountry.groupby('indicator_code').apply(lambda x: x.isnull().mean())
dfNullPercentageByIndicator

# get 1970 column id
year1970 = dfEdStatsDataCountry.columns.get_loc('1970')
# get 2016 column id
year2016 = dfEdStatsDataCountry.columns.get_loc('2016')

# we get column between 1970 and 2016 using iloc
dfNullPercentageByIndicatorPast = dfNullPercentageByIndicator.iloc[:, year1970:year2016+1]
dfNullPercentageByIndicatorPast

# we get the mean of the null values for each indicator
dfNullPercentageByIndicatorPast['mean'] = dfNullPercentageByIndicatorPast.mean(axis=1)
dfNullPercentageByIndicatorPast = dfNullPercentageByIndicatorPast.sort_values('mean', ascending=True)
dfNullPercentageByIndicatorPast

C:\Users\alexa\AppData\Local\Temp\ipykernel_8052\2263488496.py:1: DeprecationWarning:

DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.

C:\Users\alexa\AppData\Local\Temp\ipykernel_8052\2263488496.py:14: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



1970      1971      1972      1973      1974  \
indicator_code                                                           
SP.POP.TOTL           0.027650  0.027650  0.027650  0.027650  0.027650   
SP.POP.GROW           0.027650  0.027650  0.027650  0.027650  0.027650   
SE.PRM.DURS           0.073733  0.073733  0.073733  0.073733  0.073733   
SE.PRM.AGES           0.073733  0.073733  0.073733  0.073733  0.073733   
UIS.THDUR.0           0.078341  0.073733  0.073733  0.073733  0.073733   
...                        ...       ...       ...       ...       ...   
SABER.TER.GOAL6.LVL1  1.000000  1.000000  1.000000  1.000000  1.000000   
SABER.TER.GOAL6.LVL2  1.000000  1.000000  1.000000  1.000000  1.000000   
SABER.TER.GOAL6.LVL3  1.000000  1.000000  1.000000  1.000000  1.000000   
SABER.TER.GOAL4       1.000000  1.000000  1.000000  1.000000  1.000000   
SABER.TER.GOAL3       1.000000  1.000000  1.000000  1.000000  1.000000   

                          1975      1976      1977      1978      1979  ...  \
indicator_code                                                          ...   
SP.POP.TOTL           0.027650  0.027650  0.027650  0.027650  0.027650  ...   
SP.POP.GROW           0.027650  0.027650  0.027650  0.027650  0.027650  ...   
SE.PRM.DURS           0.073733  0.073733  0.073733  0.073733  0.073733  ...   
SE.PRM.AGES           0.073733  0.073733  0.073733  0.073733  0.073733  ...   
UIS.THDUR.0           0.073733  0.073733  0.073733  0.073733  0.073733  ...   
...                        ...       ...       ...       ...       ...  ...   
SABER.TER.GOAL6.LVL1  1.000000  1.000000  1.000000  1.000000  1.000000  ...   
SABER.TER.GOAL6.LVL2  1.000000  1.000000  1.000000  1.000000  1.000000  ...   
SABER.TER.GOAL6.LVL3  1.000000  1.000000  1.000000  1.000000  1.000000  ...   
SABER.TER.GOAL4       1.000000  1.000000  1.000000  1.000000  1.000000  ...   
SABER.TER.GOAL3       1.000000  1.000000  1.000000  1.000000  1.000000  ...   

                          2008      2009      2010      2011      2012  \
indicator_code                                                           
SP.POP.TOTL           0.009217  0.009217  0.009217  0.009217  0.013825   
SP.POP.GROW           0.009217  0.009217  0.009217  0.013825  0.013825   
SE.PRM.DURS           0.041475  0.041475  0.041475  0.041475  0.041475   
SE.PRM.AGES           0.041475  0.041475  0.041475  0.041475  0.041475   
UIS.THDUR.0           0.055300  0.055300  0.055300  0.055300  0.055300   
...                        ...       ...       ...       ...       ...   
SABER.TER.GOAL6.LVL1  1.000000  1.000000  1.000000  1.000000  1.000000   
SABER.TER.GOAL6.LVL2  1.000000  1.000000  1.000000  1.000000  1.000000   
SABER.TER.GOAL6.LVL3  1.000000  1.000000  1.000000  1.000000  1.000000   
SABER.TER.GOAL4       1.000000  1.000000  1.000000  1.000000  1.000000   
SABER.TER.GOAL3       1.000000  1.000000  1.000000  1.000000  1.000000   

                          2013      2014      2015      2016      mean  
indicator_code                                                          
SP.POP.TOTL           0.013825  0.032258  0.032258  0.032258  0.020590  
SP.POP.GROW           0.013825  0.032258  0.032258  0.032258  0.021179  
SE.PRM.DURS           0.041475  0.041475  0.041475  0.179724  0.055888  
SE.PRM.AGES           0.041475  0.041475  0.041475  0.179724  0.055888  
UIS.THDUR.0           0.055300  0.055300  0.055300  0.193548  0.060888  
...                        ...       ...       ...       ...       ...  
SABER.TER.GOAL6.LVL1  1.000000  1.000000  1.000000  1.000000  1.000000  
SABER.TER.GOAL6.LVL2  1.000000  1.000000  1.000000  1.000000  1.000000  
SABER.TER.GOAL6.LVL3  1.000000  1.000000  1.000000  1.000000  1.000000  
SABER.TER.GOAL4       1.000000  1.000000  1.000000  1.000000  1.000000  
SABER.TER.GOAL3       1.000000  1.000000  1.000000  1.000000  1.000000  

[3665 rows x 48 columns]

Ranking of most relevant indicators depending on null percentage for data **expected in the future**

In [210]:
# get 2020 column id
year2020 = dfNullPercentageByIndicator.columns.get_loc('2020')
# get 2100 column id
year2100 = dfNullPercentageByIndicator.columns.get_loc('2100')

# we get column between 2017 and 2100 using iloc
dfNullPercentageByIndicatorFuture = dfNullPercentageByIndicator.iloc[:, year2020:year2100+1]
dfNullPercentageByIndicatorFuture

dfNullPercentageByIndicatorFuture['mean'] = dfNullPercentageByIndicatorFuture.mean(axis=1)
dfNullPercentageByIndicatorFuture = dfNullPercentageByIndicatorFuture.sort_values('mean', ascending=True)
dfNullPercentageByIndicatorFuture

C:\Users\alexa\AppData\Local\Temp\ipykernel_8052\843556840.py:10: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



2020      2025      2030      2035      2040  \
indicator_code                                                                  
PRJ.ATT.60UP.1.MF            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.FE            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.MA            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.MF            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.NED.FE          0.235023  0.235023  0.235023  0.235023  0.235023   
...                               ...       ...       ...       ...       ...   
LO.LLECE.SCI6.3              1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.SCI6.3.FE           1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.SCI6.3.MA           1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.REA6.3.FE           1.000000  1.000000  1.000000  1.000000  1.000000   
XGDP.56.FSGOV.FDINSTADM.FFD  1.000000  1.000000  1.000000  1.000000  1.000000   

                                 2045      2050      2055      2060      2065  \
indicator_code                                                                  
PRJ.ATT.60UP.1.MF            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.FE            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.MA            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.MF            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.NED.FE          0.235023  0.235023  0.235023  0.235023  0.235023   
...                               ...       ...       ...       ...       ...   
LO.LLECE.SCI6.3              1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.SCI6.3.FE           1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.SCI6.3.MA           1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.REA6.3.FE           1.000000  1.000000  1.000000  1.000000  1.000000   
XGDP.56.FSGOV.FDINSTADM.FFD  1.000000  1.000000  1.000000  1.000000  1.000000   

                                 2070      2075      2080      2085      2090  \
indicator_code                                                                  
PRJ.ATT.60UP.1.MF            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.FE            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.MA            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.4.MF            0.235023  0.235023  0.235023  0.235023  0.235023   
PRJ.ATT.2064.NED.FE          0.235023  0.235023  0.235023  0.235023  0.235023   
...                               ...       ...       ...       ...       ...   
LO.LLECE.SCI6.3              1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.SCI6.3.FE           1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.SCI6.3.MA           1.000000  1.000000  1.000000  1.000000  1.000000   
LO.LLECE.REA6.3.FE           1.000000  1.000000  1.000000  1.000000  1.000000   
XGDP.56.FSGOV.FDINSTADM.FFD  1.000000  1.000000  1.000000  1.000000  1.000000   

                                 2095      2100      mean  
indicator_code                                             
PRJ.ATT.60UP.1.MF            0.235023  0.235023  0.235023  
PRJ.ATT.2064.4.FE            0.235023  0.235023  0.235023  
PRJ.ATT.2064.4.MA            0.235023  0.235023  0.235023  
PRJ.ATT.2064.4.MF            0.235023  0.235023  0.235023  
PRJ.ATT.2064.NED.FE          0.235023  0.235023  0.235023  
...                               ...       ...       ...  
LO.LLECE.SCI6.3              1.000000  1.000000  1.000000  
LO.LLECE.SCI6.3.FE           1.000000  1.000000  1.000000  
LO.LLECE.SCI6.3.MA           1.000000  1.000000  1.000000  
LO.LLECE.REA6.3.FE           1.000000  1.000000  1.000000  
XGDP.56.FSGOV.FDINSTADM.FFD  1.000000  1.000000  1.000000  

[3665 rows x 18 columns]

## Main indicators ranking

Population | SP.POP.1524.TO.UN

Economique | NY.GNP.PCAP.PP.CD

Education | SE.SEC.ENRR | SE.TER.ENRR

Numerique | IT.NET.USER.P2

In [211]:
numberOfIndicators= dfNullPercentageByIndicatorPast.shape[0]

# population indicator rank

populationIndicatorRankPast = dfNullPercentageByIndicatorPast.index.get_loc('SP.POP.TOTL')

populationIndicatorRankFuture = dfNullPercentageByIndicatorFuture.index.get_loc('SP.POP.TOTL')

# economic indicator rank

economicIndicatorRankPast = dfNullPercentageByIndicatorPast.index.get_loc('NY.GNP.PCAP.PP.CD')

economicIndicatorRankFuture = dfNullPercentageByIndicatorFuture.index.get_loc('NY.GNP.PCAP.PP.CD')

# eduction first indicator rank

educationIndicatorRankPast = dfNullPercentageByIndicatorPast.index.get_loc('SE.SEC.ENRR')

educationIndicatorRankFuture = dfNullPercentageByIndicatorFuture.index.get_loc('SE.SEC.ENRR')

# eduction second indicator rank

educationIndicatorRankPast2 = dfNullPercentageByIndicatorPast.index.get_loc('SE.TER.ENRR')

educationIndicatorRankFuture2 = dfNullPercentageByIndicatorFuture.index.get_loc('SE.TER.ENRR')

# numeric indicator rank

numericIndicatorRankPast = dfNullPercentageByIndicatorPast.index.get_loc('IT.NET.USER.P2')

numericIndicatorRankFuture = dfNullPercentageByIndicatorFuture.index.get_loc('IT.NET.USER.P2')

print("Past aggregated data rankings :\n")

print('Rank of SP.POP.TOTL in the list of indicators: '+ str(populationIndicatorRankPast) +'/'+ str(numberOfIndicators))

print('Rank of NY.GNP.PCAP.PP.CD in the list of indicators: '+ str(economicIndicatorRankPast) +'/'+ str(numberOfIndicators))

print('Rank of SE.SEC.ENRR in the list of indicators: '+ str(educationIndicatorRankPast) +'/'+ str(numberOfIndicators))

print('Rank of SE.TER.ENRR in the list of indicators: '+ str(educationIndicatorRankPast2) +'/'+ str(numberOfIndicators))

print('Rank of IT.NET.USER.P2 in the list of indicators: '+ str(numericIndicatorRankPast) +'/'+ str(numberOfIndicators))

print("\nFuture aggregated data rankings :\n")

print('Rank of SP.POP.TOTL in the list of indicators: '+ str(populationIndicatorRankFuture) +'/'+ str(numberOfIndicators))

print('Rank of NY.GNP.PCAP.PP.CD in the list of indicators: '+ str(economicIndicatorRankFuture) +'/'+ str(numberOfIndicators))

print('Rank of SE.SEC.ENRR in the list of indicators: '+ str(educationIndicatorRankFuture) +'/'+ str(numberOfIndicators))

print('Rank of SE.TER.ENRR in the list of indicators: '+ str(educationIndicatorRankFuture2) +'/'+ str(numberOfIndicators))

print('Rank of IT.NET.USER.P2 in the list of indicators: '+ str(numericIndicatorRankFuture) +'/'+ str(numberOfIndicators))

Past aggregated data rankings :

Rank of SP.POP.TOTL in the list of indicators: 0/3665
Rank of NY.GNP.PCAP.PP.CD in the list of indicators: 134/3665
Rank of SE.SEC.ENRR in the list of indicators: 72/3665
Rank of SE.TER.ENRR in the list of indicators: 97/3665
Rank of IT.NET.USER.P2 in the list of indicators: 132/3665

Future aggregated data rankings :

Rank of SP.POP.TOTL in the list of indicators: 1757/3665
Rank of NY.GNP.PCAP.PP.CD in the list of indicators: 3092/3665
Rank of SE.SEC.ENRR in the list of indicators: 379/3665
Rank of SE.TER.ENRR in the list of indicators: 667/3665
Rank of IT.NET.USER.P2 in the list of indicators: 2034/3665


# Choropleth maps

## Past data

Population

In [212]:
dfEdStatsDataCountry2016 = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'SP.POP.TOTL']
dfEdStatsDataCountry2016 = dfEdStatsDataCountry2016[dfEdStatsDataCountry2016['2016'].notnull()]

fig = px.choropleth(dfEdStatsDataCountry2016, locations='country_code', color='2016', hover_name='country_name', color_continuous_scale=px.colors.sequential.Plasma, title='Population in 2016', labels={'2016':'Population'})
fig.show()

> Best countries : China & India

> Interesting countries : USA, Brazil, Nigeria, Indonesia, Pakistan,... 

Economic

In [213]:
dfEdStatsDataCountry2016 = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'NY.GNP.PCAP.PP.CD']
dfEdStatsDataCountry2016 = dfEdStatsDataCountry2016[dfEdStatsDataCountry2016['2016'].notnull()]

fig = px.choropleth(dfEdStatsDataCountry2016, locations='country_code', color='2016', hover_name='country_name', color_continuous_scale=px.colors.sequential.Plasma, title='GNI per capita in 2016', labels={'2016':'GNI per capita'})
fig.show()

> Interesting countries : USA, Saudi Arabia, United Arab Emirates, Norway, Ireland, ...

Education

Secondary school enrolment rate

In [214]:
dfEdStatsDataCountry2014 = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'SE.SEC.ENRR']
dfEdStatsDataCountry2014 = dfEdStatsDataCountry2014[dfEdStatsDataCountry2014['2014'].notnull()]

fig = px.choropleth(dfEdStatsDataCountry2014, locations='country_code', color='2014', hover_name='country_name', color_continuous_scale=px.colors.sequential.Plasma, title='Secondary education enrollment in 2014', labels={'2014':'Secondary education enrollment'})
fig.show()

> Interesting countries : Australia, Thailand, Finland, Sweden, UK, Ireland, Spain, ...

Tertiary school enrolment rate

In [215]:
dfEdStatsDataCountry2014 = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'SE.TER.ENRR']
dfEdStatsDataCountry2014 = dfEdStatsDataCountry2014[dfEdStatsDataCountry2014['2014'].notnull()]

fig = px.choropleth(dfEdStatsDataCountry2014, locations='country_code', color='2014', hover_name='country_name', color_continuous_scale=px.colors.sequential.Plasma, title='Tertiary education enrollment in 2014', labels={'2014':'Secondary education enrollment'})
fig.show()

> Interesting countries : Austalia, Argentina, Chile, USA, Greece, Spain, Findland, Belarus, ...

Numeric

In [216]:
dfEdStatsDataCountry2016 = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'IT.NET.USER.P2']
dfEdStatsDataCountry2016 = dfEdStatsDataCountry2016[dfEdStatsDataCountry2016['2016'].notnull()]

fig = px.choropleth(dfEdStatsDataCountry2016, locations='country_code', color='2016', hover_name='country_name', color_continuous_scale=px.colors.sequential.Plasma, title='Internet users in 2016', labels={'2016':'Internet users'})
fig.show()

 > Interesting countries : Australia, Japan, Korea, United Arab Emirates, Norway, UK, Canada, ...

## Worth to settle in indicator

In [217]:
# create a new dataframe with all countries and all values for the indicators SP.POP.TOTL (2016), NY.GNP.PCAP.PP.CD(2016), SE.SEC.ENRR(2014), SE.TER.ENRR(2014), IT.NET.USER.P2 (2016)
dfEdStatsDataWorthToSettleIn = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'SP.POP.TOTL']
dfEdStatsDataWorthToSettleIn = dfEdStatsDataWorthToSettleIn[dfEdStatsDataWorthToSettleIn['2016'].notnull()]
dfEdStatsDataWorthToSettleIn = dfEdStatsDataWorthToSettleIn[['country_name', 'country_code', '2016']]
dfEdStatsDataWorthToSettleIn.rename(columns={'2016':'population'}, inplace=True)

gni_per_capita = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'NY.GNP.PCAP.PP.CD'][['country_code', '2016']].rename(columns={'2016':'gni_per_capita'})
dfEdStatsDataWorthToSettleIn = pd.merge(dfEdStatsDataWorthToSettleIn, gni_per_capita, on='country_code', how='left')

secondary_education_enrollment = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'SE.SEC.ENRR'][['country_code', '2014']].rename(columns={'2014':'secondary_education_enrollment'})
dfEdStatsDataWorthToSettleIn = pd.merge(dfEdStatsDataWorthToSettleIn, secondary_education_enrollment, on='country_code', how='left')

tertiary_education_enrollment = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'SE.TER.ENRR'][['country_code', '2014']].rename(columns={'2014':'tertiary_education_enrollment'})
dfEdStatsDataWorthToSettleIn = pd.merge(dfEdStatsDataWorthToSettleIn, tertiary_education_enrollment, on='country_code', how='left')

internet_users = dfEdStatsDataCountry[dfEdStatsDataCountry['indicator_code'] == 'IT.NET.USER.P2'][['country_code', '2016']].rename(columns={'2016':'internet_users'})
dfEdStatsDataWorthToSettleIn = pd.merge(dfEdStatsDataWorthToSettleIn, internet_users, on='country_code', how='left')

dfEdStatsDataWorthToSettleIn


country_name country_code  population  gni_per_capita  \
0              Afghanistan          AFG  34656032.0          1900.0   
1                  Albania          ALB   2876101.0         11670.0   
2                  Algeria          DZA  40606052.0         14420.0   
3           American Samoa          ASM     55599.0             NaN   
4                  Andorra          AND     77281.0             NaN   
..                     ...          ...         ...             ...   
205  Virgin Islands (U.S.)          VIR    102951.0             NaN   
206     West Bank and Gaza          PSE   4551566.0             NaN   
207            Yemen, Rep.          YEM  27584213.0          2500.0   
208                 Zambia          ZMB  16591390.0          3850.0   
209               Zimbabwe          ZWE  16150362.0          1810.0   

     secondary_education_enrollment  tertiary_education_enrollment  \
0                         55.656158                       8.662800   
1                         96.426086                      62.706848   
2                               NaN                      34.593811   
3                               NaN                            NaN   
4                               NaN                            NaN   
..                              ...                            ...   
205                             NaN                            NaN   
206                       82.240448                      44.006870   
207                             NaN                            NaN   
208                             NaN                            NaN   
209                             NaN                            NaN   

     internet_users  
0         10.595726  
1         66.363445  
2         42.945527  
3               NaN  
4         97.930637  
..              ...  
205       59.608316  
206       61.178385  
207       24.579208  
208       25.506579  
209       23.119989  

[210 rows x 7 columns]

In [218]:
# we fill the missing values with the mean of the column for each indicator
dfEdStatsDataWorthToSettleIn['gni_per_capita'] = dfEdStatsDataWorthToSettleIn['gni_per_capita'].fillna(dfEdStatsDataWorthToSettleIn['gni_per_capita'].mean())
dfEdStatsDataWorthToSettleIn['secondary_education_enrollment'] = dfEdStatsDataWorthToSettleIn['secondary_education_enrollment'].fillna(dfEdStatsDataWorthToSettleIn['secondary_education_enrollment'].mean())
dfEdStatsDataWorthToSettleIn['tertiary_education_enrollment'] = dfEdStatsDataWorthToSettleIn['tertiary_education_enrollment'].fillna(dfEdStatsDataWorthToSettleIn['tertiary_education_enrollment'].mean())
dfEdStatsDataWorthToSettleIn['internet_users'] = dfEdStatsDataWorthToSettleIn['internet_users'].fillna(dfEdStatsDataWorthToSettleIn['internet_users'].mean())
dfEdStatsDataWorthToSettleIn

country_name country_code  population  gni_per_capita  \
0              Afghanistan          AFG  34656032.0     1900.000000   
1                  Albania          ALB   2876101.0    11670.000000   
2                  Algeria          DZA  40606052.0    14420.000000   
3           American Samoa          ASM     55599.0    18177.570621   
4                  Andorra          AND     77281.0    18177.570621   
..                     ...          ...         ...             ...   
205  Virgin Islands (U.S.)          VIR    102951.0    18177.570621   
206     West Bank and Gaza          PSE   4551566.0    18177.570621   
207            Yemen, Rep.          YEM  27584213.0     2500.000000   
208                 Zambia          ZMB  16591390.0     3850.000000   
209               Zimbabwe          ZWE  16150362.0     1810.000000   

     secondary_education_enrollment  tertiary_education_enrollment  \
0                         55.656158                       8.662800   
1                         96.426086                      62.706848   
2                         88.508048                      34.593811   
3                         88.508048                      44.874089   
4                         88.508048                      44.874089   
..                              ...                            ...   
205                       88.508048                      44.874089   
206                       82.240448                      44.006870   
207                       88.508048                      44.874089   
208                       88.508048                      44.874089   
209                       88.508048                      44.874089   

     internet_users  
0         10.595726  
1         66.363445  
2         42.945527  
3         51.665710  
4         97.930637  
..              ...  
205       59.608316  
206       61.178385  
207       24.579208  
208       25.506579  
209       23.119989  

[210 rows x 7 columns]

In [219]:
# we create a new column with the viability score of each country based on the indicators
# we use the formula: 0.5 * population + 0.2 * (secondary_education_enrollment + tertiary_education_enrollment) + 0.3 * internet_users + 0.5 * gni_per_capita
dfEdStatsDataWorthToSettleIn['viability_score'] = 0.3 * dfEdStatsDataWorthToSettleIn['population'] + 0.4 * (dfEdStatsDataWorthToSettleIn['secondary_education_enrollment'] + dfEdStatsDataWorthToSettleIn['tertiary_education_enrollment']) + 0.5 * dfEdStatsDataWorthToSettleIn['internet_users'] + 0.5 * dfEdStatsDataWorthToSettleIn['gni_per_capita']
# we scale the viability score between 0 and 100
dfEdStatsDataWorthToSettleIn['viability_score'] = (dfEdStatsDataWorthToSettleIn['viability_score'] - dfEdStatsDataWorthToSettleIn['viability_score'].min()) / (dfEdStatsDataWorthToSettleIn['viability_score'].max() - dfEdStatsDataWorthToSettleIn['viability_score'].min()) * 100
dfEdStatsDataWorthToSettleIn.sort_values('viability_score', ascending=False, inplace=True)
dfEdStatsDataWorthToSettleIn

country_name country_code    population  gni_per_capita  \
40              China          CHN  1.378665e+09    15500.000000   
85              India          IND  1.324171e+09     6500.000000   
199     United States          USA  3.231275e+08    58700.000000   
86          Indonesia          IDN  2.611155e+08    11240.000000   
26             Brazil          BRA  2.076529e+08    14840.000000   
..                ...          ...           ...             ...   
71          Gibraltar          GIB  3.440800e+04    18177.570621   
159        San Marino          SMR  3.320300e+04    18177.570621   
121  Marshall Islands          MHL  5.306600e+04     5380.000000   
145             Palau          PLW  2.150300e+04    14870.000000   
194            Tuvalu          TUV  1.109700e+04     5930.000000   

     secondary_education_enrollment  tertiary_education_enrollment  \
40                        94.332069                      39.390388   
85                        74.275436                      25.535450   
199                       97.564690                      86.663963   
86                        82.468697                      31.102100   
26                       101.950897                      49.279839   
..                              ...                            ...   
71                        88.508048                      44.874089   
159                       88.508048                      44.874089   
121                       88.508048                      44.874089   
145                      113.597733                      44.874089   
194                       80.766640                      44.874089   

     internet_users  viability_score  
40        53.200000       100.000000  
85        29.547163        96.046281  
199       76.176737        23.443211  
86        25.366301        18.939498  
26        59.682747        15.062103  
..              ...              ...  
71        94.444472         0.003178  
159       51.665710         0.003085  
121       29.787779         0.002976  
145       51.665710         0.001839  
194       46.009152         0.000000  

[210 rows x 8 columns]

In [220]:
# we plot the choropleth map of the viability score
fig = px.choropleth(dfEdStatsDataWorthToSettleIn, locations='country_code', color='viability_score', hover_name='country_name', color_continuous_scale=px.colors.sequential.Plasma, title='Viability score', labels={'viability_score':'Viability score'})
fig.show()

## TOP 10 to settle in

In [221]:
dfEdStatsDataWorthToSettleIn.head(10)[['country_name', 'viability_score']]

country_name  viability_score
40                China       100.000000
85                India        96.046281
199       United States        23.443211
86            Indonesia        18.939498
26               Brazil        15.062103
144            Pakistan        14.012900
140             Nigeria        13.489685
15           Bangladesh        11.818407
156  Russian Federation        10.470897
124              Mexico         9.251603